File: Level - 3/Dynamic Programming/Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        //OM GAN GANAPATHAYE NAMO NAMAH 
        //JAI SHRI RAM 
        //JAI BAJRANGBALI 
        //AMME NARAYANA, DEVI NARAYANA, LAKSHMI NARAYANA, BHADRE NARAYANA
        int dp[amount+1];
        for(int i = 0; i <= amount; i++)
            dp[i] = amount + 1;
        int n = coins.size();
        for(int i = 0; i <= amount; i++)
        {
            if(i == 0)
                dp[i] = 0;
            for(int j = 0; j < n; j++)
            {
                if(coins[j] <= i)
                {
                    //put +1 only on dp[i-coins[j]] -> here only comparision occurs
                    dp[i] = min(dp[i], dp[i-coins[j]]+1);
                }
            }
        }
        if(dp[amount] > amount)
            return -1;
        if(dp[amount] == INT_MAX)
            return -1;
        return dp[amount];
    }
};

File: Blind 75 Must Do LC/Coin Change.java
// Hari

class Solution {
    public int coinChange(int[] coins, int amount) {
        Arrays.sort(coins);
        int N = coins.length;
        int[] dp = new int[amount+1];
        Arrays.fill(dp, 701);
        dp[0] = 0;
        
        // for loop for coins
        for(int i = 0; i<coins.length; i++){
            for(int j = coins[i]; j<=amount; j++){
                dp[j] = Math.min(dp[j], 1 + dp[j-coins[i]]);
            }
        }
        
        return dp[amount] == 701? -1 : dp[amount];
        
    }
}
